ldelf/ta_elf_rel.c

```diff
@@ -412,19 +412,58 @@ static void e64_process_dyn_rela(const Elf64_Sym *sym_tab, size_t num_syms,
 	*where = val;
 }
 
-static void e64_process_tls_rela(const Elf64_Sym *sym_tab, size_t num_syms,
-				 const char *str_tab, size_t str_tab_size,
-				 Elf64_Rela *rela, Elf64_Addr *where)
+static void e64_process_tls_tprel_rela(const Elf64_Sym *sym_tab,
+				       size_t num_syms, const char *str_tab,
+				       size_t str_tab_size, Elf64_Rela *rela,
+				       Elf64_Addr *where, struct ta_elf *elf)
 {
 	struct ta_elf *mod = NULL;
 	const char *name = NULL;
+	size_t sym_idx = 0;
 	vaddr_t symval = 0;
 
-	e64_get_sym_name(sym_tab, num_syms, str_tab, str_tab_size, rela, &name);
-	resolve_sym(name, &symval, &mod);
+	sym_idx = ELF64_R_SYM(rela->r_info);
+	if (sym_idx) {
+		e64_get_sym_name(sym_tab, num_syms, str_tab, str_tab_size, rela,
+				 &name);
+		resolve_sym(name, &symval, &mod);
+	} else {
+		mod = elf;
+	}
 	*where = symval + mod->tls_tcb_offs + rela->r_addend;
 }
 
+/* Helper function written in assembly due to the calling convention */
+long tlsdesc_resolve(long);
```

Comment:
The implementation of this function is more or less trivial. How about using inline assembly instead for better overview?

Reply:
Sure, I'll update.